drivers/serial/uart_nrfx_uarte.c

```diff
@@ -1040,69 +1123,66 @@ static int uarte_nrfx_poll_in(const struct device *dev, unsigned char *c)
  */
 static void uarte_nrfx_poll_out(const struct device *dev, unsigned char c)
 {
-	NRF_UARTE_Type *uarte = get_uarte_instance(dev);
 	struct uarte_nrfx_data *data = get_dev_data(dev);
-	atomic_t *lock;
-
-#ifdef CONFIG_UART_ASYNC_API
+	bool isr_mode = k_is_in_isr() || k_is_pre_kernel();
+	NRF_UARTE_Type *uarte = get_uarte_instance(dev);
+	int key;
 
-	if (data->async) {
-		while (data->async->tx_buf) {
-			/* If there is ongoing transmission, and we are in
-			 * isr, then call uarte interrupt routine, otherwise
-			 * busy wait until transmission is finished.
-			 */
-			if (k_is_in_isr()) {
+	if (isr_mode) {
+		while (1) {
+#if CONFIG_UART_ASYNC_API
+			if (data->async) {
 				uarte_nrfx_isr_async(dev);
 			}
```

Comment:
Is that still needed, actually, now that the loop waits for certain hardware event?

Reply:
we need to handle potential completion of tx transfer if we are in higher than UART irq priority context. During this completion AMOUNT register is read so it cannot be done after poll out transfer.

Reply:
Okay, I see. Non-HW byte counting would also benefit from those calls in such scenario.
But when it comes to the AMOUNT register readout, it is done in `txstopped_isr()`, so we'll have a race condition here, as the TXSTOPPED event can occur between the calls to `uarte_nrfx_isr_async()` and `is_tx_ready()`.

Reply:
yes, noticed that yesterday and working on solution.

Reply:
>  TXSTOPPED event can occur between the calls to uarte_nrfx_isr_async() and is_tx_ready().

it's not only this, if we are in lower priority interrupt then interrupt will preempt at any moment within `uarte_nrfx_isr_async` and the same other way round. Basically, we cannot have `uarte_nrfx_isr_async` there.